precompiles/act/act.go

```diff
@@ -0,0 +1,162 @@
+package act
+
+import (
+	"embed"
+	"github.com/warden-protocol/wardenprotocol/precompiles/common"
+
+	"cosmossdk.io/log"
+	storetypes "cosmossdk.io/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	ethcmn "github.com/ethereum/go-ethereum/common"
+	evmoscmn "github.com/evmos/evmos/v20/precompiles/common"
+	"github.com/evmos/evmos/v20/x/evm/core/vm"
+
+	actmodulekeeper "github.com/warden-protocol/wardenprotocol/warden/x/act/keeper"
+)
+
+var _ vm.PrecompiledContract = &Precompile{}
+
+const PrecompileAddress = "0x0000000000000000000000000000000000000901"
+
+// Embed abi json file to the executable binary. Needed when importing as dependency.
+//
+//go:embed abi.json
+var f embed.FS
+
+// Precompile defines the precompiled contract for x/act.
+type Precompile struct {
+	evmoscmn.Precompile
+	actmodulekeeper actmodulekeeper.Keeper
+	eventsRegistry  *common.EthEventsRegistry
+}
+
+// LoadABI loads the x/act ABI from the embedded abi.json file
+// for the x/act precompile.
+func LoadABI() (abi.ABI, error) {
+	return evmoscmn.LoadABI(f, "abi.json")
+}
+
+func NewPrecompile(actkeeper actmodulekeeper.Keeper, er *common.EthEventsRegistry) (*Precompile, error) {
+	abi, err := LoadABI()
+	if err != nil {
+		return nil, err
+	}
+
+	p := Precompile{
+		Precompile: evmoscmn.Precompile{
+			ABI:                  abi,
+			KvGasConfig:          storetypes.KVGasConfig(),
+			TransientKVGasConfig: storetypes.TransientGasConfig(),
+		},
+		actmodulekeeper: actkeeper,
+		eventsRegistry:  er,
+	}
+
+	return &p, nil
+}
+
+// Address implements vm.PrecompiledContract.
+func (p *Precompile) Address() ethcmn.Address {
+	return ethcmn.HexToAddress(PrecompileAddress)
+}
+
+// RequiredGas implements vm.PrecompiledContract.
+// Subtle: this method shadows the method (Precompile).RequiredGas of Precompile.Precompile.
+func (p *Precompile) RequiredGas(input []byte) uint64 {
+	methodID := input[:4]
+
+	method, err := p.MethodById(methodID)
+	if err != nil {
+		// This should never happen since this method is going to fail during Run
+		return 0
+	}
+
+	return p.Precompile.RequiredGas(input, p.IsTransaction(method.Name))
+}
+
+// Run implements vm.PrecompiledContract.
+func (p *Precompile) Run(evm *vm.EVM, contract *vm.Contract, readonly bool) (bz []byte, err error) {
+	ctx, stateDB, _, method, initialGas, args, err := p.RunSetup(evm, contract, readonly, p.IsTransaction)
+	if err != nil {
+		return nil, err
+	}
+
+	// This handles any out of gas errors that may occur during the execution of a precompile tx or query.
+	// It avoids panics and returns the out of gas error so the EVM can continue gracefully.
+	defer evmoscmn.HandleGasError(ctx, contract, initialGas, &err)()
+
+	if err := stateDB.Commit(); err != nil {
```

Comment:
do we need this commit here? i don't see it in the `staking` precompile from evmos, for example

Reply:
https://github.com/evmos/evmos/blob/main/precompiles/staking/staking.go#L97

Reply:
Hm, it's used in `gov` precompile https://github.com/evmos/evmos/blob/main/precompiles/gov/gov.go#L99

But our works without it also, so we can rm it